lib/logthrsource/logthrsourcedrv.c

```diff
@@ -0,0 +1,324 @@
+/*
+ * Copyright (c) 2018 Balabit
+ * Copyright (c) 2018 László Várady <laszlo.varady@balabit.com>
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include "logthrsourcedrv.h"
+#include "mainloop-worker.h"
+#include "messages.h"
+#include "apphook.h"
+
+typedef struct _WakeupCondition
+{
+  GMutex *lock;
+  GCond *cond;
+  gboolean awoken;
+} WakeupCondition;
+
+struct _LogThreadedSourceWorker
+{
+  LogSource super;
+  LogThreadedSourceDriver *control;
+  WakeupCondition wakeup_cond;
+  WorkerOptions options;
+
+  LogThreadedSourceWorkerRun run;
+  LogThreadedSourceWorkerRequestExit request_exit;
+  LogThreadedSourceWorkerWakeup wakeup;
+};
```

Comment:
They are used in 2 places: here and in their setter methods.

I wanted to hide the `SrcDriver --> LogSource` complexity, so my worker (`LogSource`) instance is private, you can only interact with the source driver, where setters must be used.

This can be changed later if we want to support multiple threads per source.

Reply:
Done.